### ISSUE TITLE
Geth stop importing block, possible racey mutex

### ISSUE THREAD

**Issue by Phistr90** [ethereum/go-ethereum#1936]
Hi,

i the following setup geth seems to stop importing blocks after a while:

We have a java application which works as a proxy. It gets the rpc getWork request from the ethminer forwards it to the geth node and adjusts the response. If the header hash has changed, the java application sends out a rpc call getBlockByNumber with parameter "pending" in order to get the pending block information. If the ethminer finds a solution it does the rpc call submitWork. Then this solution with reduced difficulty is send by the proxy to the node as well.

Since most solutions with reduced difficulty are not valid for the livenet it says correctly:
"Invalid header on mined block: Header's nonce is invalid (= 5228bad4da2aff10)"

After a while (on my ubuntu system after a couple of days; on users' windows machines after a day) it stops importing blocks/ sending out new work.

So to sum it up, following rpc calls are done:
*getWork
*submitWork
*getBlockByNumber("pending",...)

Somehow there is might be a racey mutex or something.

Would be nice if you could check it out.

maybe related to: https://github.com/ethereum/go-ethereum/issues/1930
or https://github.com/ethereum/go-ethereum/issues/1849

I 'll report if that happens on 1.23 too.

**Comment by Phistr90** [comment]
One windows user reported the problem again with version 1.23.
Would be nice if someone could take a look at it.

**Comment by Phistr90** [comment]
More users reporting this issue, some report a peer loss.

**Comment by karek314** [comment]
It happened to my node also, node just stuck on old block, geth didn't claim last block was invalid or something unlikely to my previous report.
Geth/v1.3.0-dev-57ab1473/linux/go1.5

**Comment by obscuren** [comment]
Attached issues do not relate.

Does anything connect using IPC?

**Comment by Phistr90** [comment]
"Does anything connect using IPC?"

The java client does only the rpc calls listed above.

**Comment by obscuren** [comment]
I understand, but does it make use of HTTP or IPC?

**Comment by Phistr90** [comment]
http
